### PR TITLE
XInput Extension Revamp

### LINKED ISSUE #900: gamepad_button_check/gamepad_button_check_pressed does not function properly.
Using gamepad_button_check/_pressed in place of keyboard_check/_pressed does not work as expected.

I am working on a menu, and this causes the menu to move rapidly when the gamepad is pressed. 

`if keyboard_check_pressed(vk_left) or gamepad_button_check_pressed(0, gp_padl) { move left once }`

Keyboard works as expected.

This:

`if keyboard_check(vk_left) 
or gamepad_button_check(0, gp_padl)
    {
        leftTime+=1
    }
`

Does not work as expected when gamepad is pressed. It actually appears to do nothing at all.

### LINKED ISSUE #1000: gamepad_get_device_count not working correctly
gamepad_get_device_count returns 32437344 with no controllers plugged in, and returns 32437345 with one plugged in.

This is quite wrong, and can cause game hanging with code such as:
`for (i=0; i<gamepad_get_device_count(); i+=1)`

### PR DESCRIPTION
**Pull request by RobertBColton**
Alright, this one has been on the conveyor belt for a while but my previous attempts (such as #966) have gone stale. There's numerous issues with the original implementation of XInput gamepads that I wrote. These issues range from simple typos to querying the gamepad state for each button check instead of once per frame. This pull request attempts to address the lot of them.

So the biggest difference here by far is the introduction of an actual `Gamepad` struct which holds various XInput state structures associated with each device. The extension also uses an init function to register an update hook for itself where it processes the gamepads and caches their state each frame. This is potentially more performant because the gamepad doesn't have to be queried multiple times a frame for each button check. This also allows us to cache the previous state of the gamepad which we now use to properly detect when buttons are pressed and released instead of just checking if the button is down.

### Summary of Changes
* Removed `gamepad_get_max_device_count` which is not a GMS function because it's not really useful and I also want to avoid making the API confusing.
* Implemented `gamepad_set_axis_deadzone` which was previously missing to make it easy to configure a deadzone for the thumbsticks of a gamepad. This way users don't have to write complicated logic to do it themselves too.
* Added `gamepad_set_color` which was previously missing along with a `gamepad_set_colour` overload. The function is a noop with the XInput extension because it's not supported and it will display an appropriate non-fatal error message in debug mode to warn of such. Only PlayStation controllers can have the color of their indicator LED changed.
* Added `enigma::extension_xinput_init` for initializing the XInput extension, getting the initial state of gamepads for the create/start events, and registering an update hook to regularly query the gamepad state.
* Updated the copyright year by adding 2018 since significant changes are being made by me in the year of this pull request. Also fixed a typo containing random braces inside the copyright license.
* Moved `digitalButtons` array to an anonymous namespace to avoid potential linking conflicts. I also changed the mapping of `gp_face2` to match up with B instead of X and vice versa for `gp_face3` because I had them backwards. This was discovered by @SuperRiderTH on #966.
* Created a `button_value` helper function in an anonymous namespace to assist with obtaining the value of analog and digital buttons.
 * Added a `Gamepad` struct to an anonymous namespace that holds other XInput structures containing the current state of a device.
* Added `process_gamepads` update hook to an anonymous namespace which the extension uses to query the gamepad state each frame.
* Reduced the `gamepad_is_connected` function to just check if the device id is between 0 and 3 inclusive and whether the controller was connected when it was last queried by the update hook.
* Reduced the `gamepad_get_device_count` function to just return the max XInput users constant, which is 4, so that it's the same as GMS. This resolves #1000 reported in the current master version which is caused by the fact I wasn't initializing the local int I was using to count the connected devices.
* Updated `gamepad_get_description` to offer more unique descriptions for other subtypes of XInput controllers.
* Updated `gamepad_get_battery_type` and `gamepad_get_battery_charge` to use the battery state that was cached by the update hook instead of querying it on the fly.
* Changed `gamepad_get_button_threshold` to actually return the button threshold used by the given gamepad.
* Changed `gamepad_set_button_threshold` to actually set/store the button threshold of the gamepad.
* Changed `gamepad_axis_count` to use the cached capabilities information of the gamepad to count the number of axes. This reports 4 for my GameStop-branded (model BB-070) standard Xbox 360 controller in GMS and this pull request.
* Changed `gamepad_axis_value` to implement the deadzone behavior as well as invert the vertical axes. The vertical axes are obviously inverted by GMS to match the coordinate system of its graphics system where the screen origin is top-left, so we should do the same.
* Changed `gamepad_button_check` to return true for digital buttons that are down and analogue buttons that have a value beyond the current button threshold of the controller.
* Changed `gamepad_button_check_pressed` to return true if the button is down but it was not down during the previous frame. The function can be safely used with analogue buttons now as well because it utilizes the same button threshold as `gamepad_button_check`. This resolves #900 by looking at the state of the gamepad in the last frame instead of just checking if the button is currently down.
* Changed `gamepad_button_check_released` to return true if the button is not down but it was down during the previous frame. The function can be safely used with analogue buttons now as well because it utilizes the same button threshold as `gamepad_button_check`. This resolves #900 by looking at the state of the gamepad in the last frame instead of just checking if the button is currently not down.
* Changed `gamepad_button_count` to use the capabilities information to count the number of buttons on the gamepad. This reports 16 for my GameStop-branded (model BB-070) standard Xbox 360 controller in GMS and this pull request.
* Changed `gamepad_button_value` to return the raw value of an analog or digital gamepad button returned by XInput normalized in the range 0-1, where 0 is no pressure and 1 is full pressure. I also changed the mapping of `gp_shoulderl` and `gp_shoulderr` to `gp_shoulderlb` and `gp_shoulderrb` respectively so that the function refers to the shoulder triggers and not the shoulder buttons which resolves the issue reported in #1035.
* Reformatted the extension to use two spaces for indentation, consistent with ENIGMA's style guidelines, instead of tabs. This also removes a lot of excess whitespace that is not needed.

### CODE REVIEW

**Comment by RobertBColton** [comment]
Alright @JoshDreamland this one is ready for review when you get a chance, thanks!

File: ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp [hunk 1]
```diff
@@ -17,238 +16,321 @@
...
+        return "Xbox 360 Controller (XInput GUITAR BASS)";
+#endif
+      case XINPUT_DEVSUBTYPE_DRUM_KIT:
+        return "Xbox 360 Controller (XInput DRUM KIT)";
+#ifdef XINPUT_DEVSUBTYPE_ARCADE_PAD
+      case XINPUT_DEVSUBTYPE_ARCADE_PAD:
+        return "Xbox 360 Controller (XInput ARCADE PAD)";
+#endif
+    }
+    return "Xbox 360 Controller (XInput UNKNOWN)";
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp]: What about non XBox controllers? Does XInput just not supply a mechanism for these?

File: ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp [hunk 2]
```diff
@@ -17,238 +16,321 @@
...
+  }
+  const float deadzone = gamepads[device].axis_deadzone;
+  if (ret >= deadzone) {
+    ret = (ret - deadzone) / (1.0f - deadzone);
+  } else if (ret <= -deadzone) {
+    ret = (ret + deadzone) / (1.0f - deadzone);
+  } else {
+    return 0;
+  }
+  return (vertical ? -ret : ret);
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp]: Is this seriously the GM behavior? Best not to lie to people about what axis the joystick is reporting. Let the developer decide that the axis bound to up/down should be inverted.

File: ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp (hunk 1 above)
Comment (RobertBColton) [reply ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp]: I don't think so, I think you have to install a special driver for a Sixaxis/DualShock to be recognized by XInput. That's probably why YoYo also loads 8 DirectInput gamepads.
https://docs2.yoyogames.com/source/_build/3_scripting/4_gml_reference/controls/gamepad%20input/index.html

>Windows is fully supported with up to a maximum of 12 connected devices permitted at once (numbered from 0 to 11, with 0 - 3 being XInput devices and 4 - 11 being DirectInput).

That said, I've heard Microsoft has been very good at getting XInput to recognize a lot of third party controllers, like Logitech. Anyway, we can come back to this later, but here's some more info about DirectInput interoperability with XInput:
https://docs.microsoft.com/en-us/windows/desktop/xinput/xinput-and-directinput

**Pull request merged**